node/data_node.go

```diff
@@ -0,0 +1,341 @@
+
+	node.catchupBlockAuth = blockAuthenticatorImpl{Ledger: node.ledger, AsyncVoteVerifier: agreement.MakeAsyncVoteVerifier(node.lowPriorityCryptoVerificationPool)}
+	node.catchupService = catchup.MakeService(node.log, node.config, p2pNode, node.ledger, node.catchupBlockAuth, make(chan catchup.PendingUnmatchedCertificate), node.lowPriorityCryptoVerificationPool)
+
+	// Initialize sync round to the next round so that nothing falls out of the cache on Start
+	err = node.SetSyncRound(uint64(node.Ledger().NextRound()))
+	if err != nil {
+		log.Errorf("unable to set sync round to Ledger.NextRound %v", err)
+		return nil, err
+	}
+
+	catchpointCatchupState, err := node.ledger.GetCatchpointCatchupState(context.Background())
+	if err != nil {
+		log.Errorf("unable to determine catchpoint catchup state: %v", err)
+		return nil, err
+	}
+	if catchpointCatchupState != ledger.CatchpointCatchupStateInactive {
+		accessor := ledger.MakeCatchpointCatchupAccessor(node.ledger.Ledger, node.log)
+		node.catchpointCatchupService, err = catchup.MakeResumedCatchpointCatchupService(context.Background(), node, node.log, node.net, accessor, node.config)
+		if err != nil {
+			log.Errorf("unable to create catchpoint catchup service: %v", err)
+			return nil, err
+		}
+		node.log.Infof("resuming catchpoint catchup from state %d", catchpointCatchupState)
+	}
+
+	node.tracer = messagetracer.NewTracer(log).Init(cfg)
+
+	stateProofPathname := filepath.Join(genesisDir, config.StateProofFileName)
+	stateProofAccess, err := db.MakeAccessor(stateProofPathname, false, false)
+	if err != nil {
+		log.Errorf("Cannot load state proof data: %v", err)
+		return nil, err
+	}
+	node.stateProofWorker = stateproof.NewWorker(stateProofAccess, node.log, node.accountManager, node.ledger.Ledger, node.net, node)
+
+	return node, err
+}
+
+// Start the node: connect to peers while obtaining a lock. Doesn't wait for initial sync.
+func (node *AlgorandDataNode) Start() {
+	node.mu.Lock()
+	defer node.mu.Unlock()
+
+	// Set up a context we can use to cancel goroutines on Stop()
+	node.ctx, node.cancelCtx = context.WithCancel(context.Background())
+
+	// The start network is being called only after the various services start up.
+	// We want to do so in order to let the services register their callbacks with the
+	// network package before any connections are being made.
+	startNetwork := func() {
+		if !node.config.DisableNetworking {
+			// start accepting connections
+			node.net.Start()
+			node.config.NetAddress, _ = node.net.Address()
+		}
+	}
+
+	if node.catchpointCatchupService != nil {
+		startNetwork()
+		node.catchpointCatchupService.Start(node.ctx)
+	} else {
+		node.catchupService.Start()
+		node.blockService.Start()
+		node.ledgerService.Start()
+		node.stateProofWorker.Start()
+		startNetwork()
+
+		node.startMonitoringRoutines()
+	}
+
+}
+
+// startMonitoringRoutines starts the internal monitoring routines used by the node.
+func (node *AlgorandDataNode) startMonitoringRoutines() {
```

Comment:
I would consider making this a pure function with 4 arguments to avoid code duplication:
```suggestion
func startMonitoringRoutines(ctx Context.context, enableUsageLog bool, wg sync.WaitGroup, log logging.Logging) {
```

On the other hand... it looks like there could be a whole lot of this... so maybe there is a better pattern?